protofsm/state_machine.go

```diff
@@ -0,0 +1,428 @@
+	newStateEvents *fn.EventDistributor[State[Event, Env]]
+
+	stateQuery chan stateQuery[Event, Env]
+
+	// TODO(roasbeef): also use that context guard here?
+}
+
+// NewStateMachine creates a new state machine given a set of daemon adapters,
+// an initial state, and an environment.
+func NewStateMachine[Event any, Env Environment](adapters DaemonAdapters,
+	initialState State[Event, Env],
+	env Env) StateMachine[Event, Env] {
+
+	return StateMachine[Event, Env]{
+		daemon:         adapters,
+		events:         make(chan Event),
+		currentState:   initialState,
+		stateQuery:     make(chan stateQuery[Event, Env]),
+		quit:           make(chan struct{}),
+		env:            env,
+		newStateEvents: fn.NewEventDistributor[State[Event, Env]](),
+	}
+}
+
+// Start starts the state machine. This will spawn a goroutine that will drive
+// the state machine to completion.
+func (s *StateMachine[Event, Env]) Start() {
+	s.wg.Add(1)
+	go s.driveMachine()
+}
+
+// Stop stops the state machine. This will block until the state machine has
+// reached a stopping point.
+func (s *StateMachine[Event, Env]) Stop() {
+	close(s.quit)
+	s.wg.Wait()
+}
+
+// SendEvent sends a new event to the state machine.
+//
+// TODO(roasbeef): bool if processed?
+func (s *StateMachine[Event, Env]) SendEvent(event Event) {
+	select {
+	case s.events <- event:
+	case <-s.quit:
+		return
+	}
+}
+
+// CurrentState returns the current state of the state machine.
+func (s *StateMachine[Event, Env]) CurrentState() (State[Event, Env], error) {
+	query := stateQuery[Event, Env]{
+		CurrentState: make(chan State[Event, Env], 1),
+	}
+
+	if !fn.SendOrQuit(s.stateQuery, query, s.quit) {
+		return nil, fmt.Errorf("state machine is shutting down")
+	}
+
+	// TODO(roasbeef): channel sent on for each state transition?
+	// subscribers?
+
+	return fn.RecvOrTimeout(query.CurrentState, time.Second)
+}
+
+// StateSubscriber represents an active subscription to be notified of new
+// state transitions.
+type StateSubscriber[E any, F Environment] *fn.EventReceiver[State[E, F]]
+
+// RegisterStateEvents registers a new event listener that will be notified of
+// new state transitions.
+func (s *StateMachine[Event, Env]) RegisterStateEvents() StateSubscriber[Event, Env] {
+	subscriber := fn.NewEventReceiver[State[Event, Env]](10)
+
+	// TODO(roasbeef): instead give the state and the input event?
+
+	s.newStateEvents.RegisterSubscriber(subscriber)
+
+	return subscriber
+}
+
+// RemoveStateSub removes the target state subscriber from the set of active
+// subscribers.
+func (s *StateMachine[Event, Env]) RemoveStateSub(sub StateSubscriber[Event, Env]) {
+	s.newStateEvents.RemoveSubscriber(sub)
+}
+
+// executeDaemonEvent executes a daemon event, which is a special type of event
+// that can be emitted as part of the state transition function of the state
+// machine. An error is returned if the type of event is unknown.
+func (s *StateMachine[Event, Env]) executeDaemonEvent(event DaemonEvent) error {
```

Comment:
feels like it's leaking the implementation details from other subsystems

Reply:
I think you'll have a better idea of the interaction once the new co-op close stuff is up, but the general idea is that:
  * All the state machine state transitions are pure functions 
  * They emit events for the executor (prob should rename this struct slightly) to apply themselves 
  * Something needs to be aware of the boundary between the pure state machine, and the daemon execution env it runs in
  * This thing handles that role of knowing all the global I/O or daemon actions to execute itself, and potentially emit an event back into the state machine (post execution hook) 

Otherwise, what do you think should be handling the I/O between the daemon and the state machine?

Reply:
Maybe it could hidden behind `currentState.ProcessEvent`? Since it generates the `transition`, it might as well process it based on the new `transition`, like broadcast or send message.

Reply:
> Maybe it could hidden behind currentState.ProcessEvent? Since it generates the transition, it might as well process it based on the new transition, like broadcast or send message.

So the idea is that the actual state transitions never need to concern themselves with any of these details. They just emit the event, then wait for w/e new event to be sent in. There's no leakage of implementation details at this `StateMachine` level, as we'll pass in a concrete implementation based on lnd later, here's an idea of what that looks like: https://github.com/lightningnetwork/lnd/pull/8453/commits/ce75ef8861923696bc9edc2deb01093b11850b6a.


This is to be considered universal, just like the `POSIX` interface we all know and love today. In this case, our processes re these FSMs, and the syscalls ways to interact with the chain or daemon.